### PR TITLE
feat: add transaction propagation

### PR DESCRIPTION
**Pull request by beeme1mr**
Signed-off-by: Michael Beemer <beeme1mr@users.noreply.github.com>

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -284,3 +284,22 @@ export interface Hook<T extends FlagValue = FlagValue> {
    */
   finally?(hookContext: Readonly<HookContext<T>>, hookHints?: HookHints): Promise<void> | void;
 }
+
+/**
+ * Transaction context is a mechanism for adding transaction specific context that
+ * is merged with evaluation context prior to flag evaluation. Examples of potential
+ * transaction specific context include: a user id, user agent, or IP address.
+ */
+export type TransactionContext = EvaluationContext;
```
Comment (beeme1mr) [inline src/types.ts:293]: @toddbaert what do you think about this? We could use evaluation context but having a transaction concept type seemed nice.

File: src/types.ts [hunk 2]
```diff
@@ -520,3 +527,56 @@ export interface Hook<T extends FlagValue = FlagValue> {
...
+
+export interface TransactionContextPropagator {
+  /**
+   * EXPERIMENTAL: Transaction context propagation is experimental and subject to change.
+   * The OpenFeature Enhancement Proposal regarding transaction context can be found [here](https://github.com/open-feature/ofep/pull/32).
+   *
+   * Returns the currently defined transaction context using the registered transaction
+   * context propagator.
+   *
+   * @returns {TransactionContext} The currently defined transaction context
```
Comment (toddbaert) [inline src/types.ts:562]:
```suggestion
   * @returns {TransactionContext} The current transaction context
```

File: src/open-feature.ts [hunk 3]
```diff
@@ -78,6 +91,23 @@ class OpenFeatureAPI implements GlobalApi {
...
+  setTransactionContext<R>(
+    transactionContext: TransactionContext,
+    callback: (...args: unknown[]) => R,
+    ...args: unknown[]
+  ): void {
+    this._transactionContextPropagator.setTransactionContext(transactionContext, callback, ...args);
+  }
+
+  getTransactionContext(): TransactionContext {
+    return this._transactionContextPropagator.getTransactionContext();
```
Comment (toddbaert) [inline src/open-feature.ts:109]: We should think about whether it's better to throw (and therefore default) if something goes wrong here, or just ignore the transaction context.

For example, if `getTransactionContext` is not a function, that will be an exception, which will cause a default.... or we could check that is a function, and return undefined. I think throwing/defaulting is PROBABLY better, but I'm not sure.

We could also do some validation at registration time like we do for the logger.

File: src/client.ts [hunk 4]
```diff
@@ -194,6 +194,7 @@ export class OpenFeatureClient implements Client {
     // merge global and client contexts
     const mergedContext = {
       ...OpenFeature.getContext(),
+      ...OpenFeature.getTransactionContext(),
```
Comment (toddbaert) [inline src/client.ts:197]: This needs to be in the try/catch now I think. `getTransactionContext()` runs user code.

**Review by toddbaert — CHANGES_REQUESTED** [review]
https://github.com/open-feature/js-sdk/pull/212/files#r984831394

**Review by toddbaert — APPROVED** [review]

File: src/open-feature.ts (hunk 3 above)
Comment (toddbaert) [reply src/open-feature.ts]: fixed

File: src/client.ts (hunk 4 above)
Comment (toddbaert) [reply src/client.ts:197]: fixed

**Pull request merged**